### ISSUE TITLE
[vim9script] “Missing return statement” raised when function never returns

### ISSUE THREAD

**Issue by lifepillar** [vim/vim#12262]
If you source this script:

```vim
vim9script

def F(): number
  if false
    return 0
  endif

  throw printf('Error')
enddef

defcompile
```

you will get `Missing return statement` (at least up to Vim 9.0.1454). Although the error is technically correct and easily fixed by adding a moot `return` statement after the `throw` statement (but shouldn't that fail to compile with “unreachable code” then?), maybe when the control never returns to the caller as in the example above, that could be accepted as valid syntax.

**Comment by lacygoill** [comment]
The new error is confusing:
```vim
vim9script
def Func(): number
    throw 'error'
    return 0
enddef
defcompile
```
    E1095: Unreachable code after :return

There is no unreachable code after `:return`, since there is no code at all afterward.
IMO, a better error would be:

    E1095: Unreachable code after :throw

**Comment by brammool** [comment]
> The new error is confusing:
> ```vim
> vim9script
> def Func(): number
>     throw 'error'
>     return 0
> enddef
> defcompile
> ```
>     E1095: Unreachable code after :return
> 
> There is no unreachable code after `:return`, since there is no code at all afterward.
> IMO, a better error would be:
> 
>     E1095: Unreachable code after :throw

Yes, I also noticed that while testing the change.  Fixed with patch 9.0.1558.

There probably are some situations where a "throw" is not handled
properly, e.g. when in every block of if/elseif.  This requires coming
up with a few examples (which can then be used in a test).  Hmm,
thinking about it, this should be one:

	if condition
	  throw 'true'
	elseif other
	  throw 'maybe'
	else
	  throw 'no'
	endif
	echo 'not reached'

Some of the "throw" commands might be "return", this needs a generic
solution.

-- 
hundred-and-one symptoms of being an internet addict:
34. You laugh at people with a 10 Mbit connection.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
